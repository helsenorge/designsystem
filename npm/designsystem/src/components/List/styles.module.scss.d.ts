export type Styles = {
  list: string;
  list__item: string;
  'list--alphabetical': string;
  'list--bullet': string;
  'list--dashed': string;
  'list--margin': string;
  'list--numbered': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
