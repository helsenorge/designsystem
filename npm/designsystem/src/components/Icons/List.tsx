import React from 'react';
import { getIcon, SvgPathProps } from './Icon';

const List: React.FC<SvgPathProps> = ({ size, isHovered }: SvgPathProps): JSX.Element => {
  const normal = (
    <path
      fillRule={'evenodd'}
      d="M9.41 32.62a1.079 1.079 0 100 2.159 1.079 1.079 0 000-2.158zm0-9.697a1.079 1.079 0 100 2.158 1.079 1.079 0 000-2.158zm0-9.702a1.079 1.079 0 100 2.158 1.079 1.079 0 000-2.158zm5.817 21.13h24.666V33.05H15.227v1.3zm0-9.7h24.666v-1.3H15.227v1.3zm0-9.697h24.665v-1.301H15.226v1.3z"
    />
  );

  const normalHover = (
    <path
      fillRule={'evenodd'}
      d="M9.41 34.62a1.079 1.079 0 100 2.159 1.079 1.079 0 000-2.158zm0-11.697a1.079 1.079 0 100 2.158 1.079 1.079 0 000-2.158zm0-11.702a1.079 1.079 0 100 2.158 1.079 1.079 0 000-2.158zm5.817 25.13h24.666V35.05H15.227v1.3zm0-11.7h24.666v-1.3H15.227v1.3zm0-11.697h24.665v-1.301H15.226v1.3z"
    />
  );

  const small = (
    <path
      fillRule={'evenodd'}
      d="M11.25 31.336a1.137 1.137 0 100 2.275 1.137 1.137 0 000-2.275zm0-8.474a1.137 1.137 0 100 2.275 1.137 1.137 0 000-2.275zm0-8.477a1.137 1.137 0 100 2.275 1.137 1.137 0 000-2.275zm5.085 19.226h21.552v-2.275H16.335v2.275zm0-8.474h21.552v-2.275H16.335v2.275zm0-8.475h21.552v-2.275H16.335v2.275z"
    />
  );

  const smallHover = (
    <path
      fillRule={'evenodd'}
      d="M11.25 33.863a1.137 1.137 0 100 2.275 1.137 1.137 0 000-2.275zm0-11.001a1.137 1.137 0 100 2.275 1.137 1.137 0 000-2.275zm0-11a1.137 1.137 0 100 2.274 1.137 1.137 0 000-2.275zm5.084 24.277h21.553v-2.275H16.334v2.275zm0-11.002h21.553v-2.275H16.334v2.275zm0-10.997h21.553v-2.275H16.334v2.275z"
    />
  );
  return getIcon(size, isHovered, normal, normalHover, small, smallHover);
};

export default List;
