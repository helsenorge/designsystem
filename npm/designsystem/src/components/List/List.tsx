import React from 'react';

import classNames from 'classnames';

import { AnalyticsId } from '../../constants';

import styles from './styles.module.scss';

export type ListVariant = 'bullet' | 'dashed' | 'numbered' | 'alphabetical';

export interface ItemProps {
  children: React.ReactNode;
}

type ItemType = React.FC<ItemProps>;

const Item: ItemType = ({ children }: ItemProps) => <li className={styles.list__item}>{children}</li>;

export interface ListProps {
  /** List contents */
  children: React.ReactNode;
  /** Changes the visual representation of the list. */
  variant?: ListVariant;
  /** Adds margin above/below list */
  margin?: boolean;
  /** Adds className to list element */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export interface ListCompound extends React.FC<ListProps> {
  Item: ItemType;
}

const List: ListCompound = ({ children, className, testId, variant = 'bullet', margin }: ListProps) => {
  const CustomTag = ['numbered', 'alphabetical'].includes(variant) ? 'ol' : 'ul';

  const listClasses = classNames(styles.list, styles[`list--${variant}`], margin && styles['list--margin'], className);

  return (
    <CustomTag className={listClasses} data-testid={testId} data-analyticsid={AnalyticsId.List}>
      {children}
    </CustomTag>
  );
};

List.Item = Item;

export default List;
