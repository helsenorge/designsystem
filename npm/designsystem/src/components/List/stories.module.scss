.list-anchorlink {
  li {
    button {
      vertical-align: top;
    }
  }
}
