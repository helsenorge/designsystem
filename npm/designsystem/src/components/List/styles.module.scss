@import '../../scss/spacers';

.list {
  $list: &;

  margin-top: 0;
  margin-bottom: 0;

  &--margin {
    margin-top: getSpacer(s);
    margin-bottom: getSpacer(s);
  }

  &--bullet {
    list-style-type: disc;

    > #{$list}__item {
      > #{$list}--bullet {
        list-style-type: circle;

        > #{$list}__item {
          > #{$list}--bullet {
            list-style-type: '⬥ ';

            > #{$list}__item {
              &::marker {
                font-size: 1rem;
              }
            }
          }
        }
      }
    }
  }

  &--dashed {
    list-style-type: '– ';
  }

  &--numbered {
    list-style-type: decimal;
  }

  &--alphabetical {
    list-style-type: lower-alpha;
  }

  &__item {
    > #{$list} {
      margin-top: getSpacer(2xs);
    }
  }

  &__item:not(:last-child) {
    margin-bottom: getSpacer(2xs);
  }
}
