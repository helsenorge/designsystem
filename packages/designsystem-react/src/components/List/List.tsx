import React from 'react';
import styled from 'styled-components';
import {getColor} from '../../theme/currys/color';

const StyledListItem = styled('li')`
  border-bottom: 1px solid ${getColor('neutral', 600)};
`;

const StyledList = styled('ul')`
  width: 100%;
  list-style: none;
  padding: 0 1rem 0 0.5rem;
`;

interface ListProps {
  children: React.ReactNode[];
  className?: string;
}

function List(props: ListProps) {
  const {children, className} = props;
  return (
    <StyledList className={className}>
      {children.map((child: React.ReactNode, index: number) => {
        return <StyledListItem key={index}>{child}</StyledListItem>;
      })}
    </StyledList>
  );
}

export default List;
