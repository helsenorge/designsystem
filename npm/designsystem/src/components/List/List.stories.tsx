import React from 'react';

import { Title as DocsTitle, Subtitle, Description, Primary, ArgsTable, PRIMARY_STORY } from '@storybook/addon-docs';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import List from './List';
import { mediumLoremText } from '../../utils/loremtext';
import GridExample from '../GridExample';

export default {
  title: 'Components/List',
  component: List,
  parameters: {
    docs: {
      description: {
        component: 'Strukturert visning av punktvise data',
      },
      page: () => (
        <>
          <DocsTitle />
          <Subtitle />
          <Description />
          <Primary />
          <ArgsTable story={PRIMARY_STORY} />
        </>
      ),
    },
  },
  argTypes: {
    variant: {
      control: 'select',
      options: ['bullet', 'dashed', 'numbered', 'alphabetical'],
      defaultValue: 'bullet',
    },
    margin: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof List>;

export const Default: ComponentStory<typeof List> = args => (
  <GridExample>
    <List {...args}>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>
        {mediumLoremText}
        <List variant={args.variant}>
          <List.Item>{mediumLoremText}</List.Item>
          <List.Item>{mediumLoremText}</List.Item>
          <List.Item>{mediumLoremText}</List.Item>
          <List.Item>
            {mediumLoremText}
            <List variant={args.variant}>
              <List.Item>{mediumLoremText}</List.Item>
              <List.Item>{mediumLoremText}</List.Item>
              <List.Item>{mediumLoremText}</List.Item>
              <List.Item>
                {mediumLoremText}
                <List variant={args.variant}>
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>
                    {mediumLoremText}
                    <List variant={args.variant}>
                      <List.Item>{mediumLoremText}</List.Item>
                      <List.Item>{mediumLoremText}</List.Item>
                      <List.Item>{mediumLoremText}</List.Item>
                      <List.Item>{mediumLoremText}</List.Item>
                    </List>
                  </List.Item>
                </List>
              </List.Item>
            </List>
          </List.Item>
        </List>
      </List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
      <List.Item>{mediumLoremText}</List.Item>
    </List>
  </GridExample>
);

export const Bullet: ComponentStory<typeof List> = () => (
  <GridExample>
    <List variant="bullet">
      <List.Item>
        {mediumLoremText}
        <List variant="bullet">
          <List.Item>
            {mediumLoremText}
            <List variant="bullet">
              <List.Item>
                {mediumLoremText}
                <List variant="bullet">
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>{mediumLoremText}</List.Item>
                </List>
              </List.Item>
              <List.Item>{mediumLoremText}</List.Item>
            </List>
          </List.Item>
          <List.Item>{mediumLoremText}</List.Item>
        </List>
      </List.Item>
      <List.Item>{mediumLoremText}</List.Item>
    </List>
  </GridExample>
);

export const Dashed: ComponentStory<typeof List> = () => (
  <GridExample>
    <List variant="dashed">
      <List.Item>
        {mediumLoremText}
        <List variant="dashed">
          <List.Item>
            {mediumLoremText}
            <List variant="dashed">
              <List.Item>
                {mediumLoremText}
                <List variant="dashed">
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>{mediumLoremText}</List.Item>
                </List>
              </List.Item>
              <List.Item>{mediumLoremText}</List.Item>
            </List>
          </List.Item>
          <List.Item>{mediumLoremText}</List.Item>
        </List>
      </List.Item>
      <List.Item>{mediumLoremText}</List.Item>
    </List>
  </GridExample>
);

export const Numbered: ComponentStory<typeof List> = () => (
  <GridExample>
    <List variant="numbered">
      <List.Item>
        {mediumLoremText}
        <List variant="numbered">
          <List.Item>
            {mediumLoremText}
            <List variant="numbered">
              <List.Item>
                {mediumLoremText}
                <List variant="numbered">
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>{mediumLoremText}</List.Item>
                </List>
              </List.Item>
              <List.Item>{mediumLoremText}</List.Item>
            </List>
          </List.Item>
          <List.Item>{mediumLoremText}</List.Item>
        </List>
      </List.Item>
      <List.Item>{mediumLoremText}</List.Item>
    </List>
  </GridExample>
);

export const Nested: ComponentStory<typeof List> = () => (
  <GridExample>
    <List variant="numbered">
      <List.Item>
        {mediumLoremText}
        <List variant="bullet">
          <List.Item>
            {mediumLoremText}
            <List variant="alphabetical">
              <List.Item>
                {mediumLoremText}
                <List variant="dashed">
                  <List.Item>{mediumLoremText}</List.Item>
                  <List.Item>{mediumLoremText}</List.Item>
                </List>
              </List.Item>
              <List.Item>{mediumLoremText}</List.Item>
            </List>
          </List.Item>
          <List.Item>{mediumLoremText}</List.Item>
        </List>
      </List.Item>
      <List.Item>{mediumLoremText}</List.Item>
    </List>
  </GridExample>
);
