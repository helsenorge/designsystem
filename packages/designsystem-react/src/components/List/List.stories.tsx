import React from 'react';
import {withKnobs} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import List from './List';
import {withA11y} from '@storybook/addon-a11y';

const stories = storiesOf('List', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <List>
    <h1>Tjeneste 1</h1>
    <h1>Tjeneste 2</h1>
    <h1>Tjeneste 3</h1>
    <h1>Tjeneste 4</h1>
  </List>
));
