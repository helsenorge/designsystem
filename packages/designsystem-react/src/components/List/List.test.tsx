import React from 'react';
import {render} from '@testing-library/react';
import List from './List';

test('displays a list of 4 elements', (): void => {
  const {container, getAllByRole} = render(
    <List>
      <h1>Tjeneste 1</h1>
      <h1>Tjeneste 2</h1>
      <h1>Tjeneste 3</h1>
      <h1>Tjeneste 4</h1>
    </List>,
  );
  expect(getAllByRole('heading').length).toEqual(4);
  expect(container).toMatchSnapshot();
});
