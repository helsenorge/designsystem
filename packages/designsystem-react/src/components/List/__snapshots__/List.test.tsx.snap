// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a list of 4 elements 1`] = `
<div>
  <ul
    class="sc-AxirZ fOOpMY"
  >
    <li
      class="sc-AxjAm guaFVz"
    >
      <h1>
        Tjeneste 1
      </h1>
    </li>
    <li
      class="sc-AxjAm guaFVz"
    >
      <h1>
        Tjeneste 2
      </h1>
    </li>
    <li
      class="sc-AxjAm guaFVz"
    >
      <h1>
        Tjeneste 3
      </h1>
    </li>
    <li
      class="sc-AxjAm guaFVz"
    >
      <h1>
        Tjeneste 4
      </h1>
    </li>
  </ul>
</div>
`;
